File: scripts/run_chain_tests.py
#!/usr/bin/env python3

import argparse
import os
import psutil
import signal
import subprocess
import time

def run_process(args):
    return subprocess.check_output(args, stderr=subprocess.STDOUT,
                                   encoding="utf-8")

def run_chain_test(chain) -> bool:
    """
    Run tests against a specific chain, returning a boolean for
    whether the tests passed or not.
    """
    run_process(["yarn"])

    if chain == "polkadot":
        url = "wss://rpc.polkadot.io"
        chain = "polkadot"
    elif chain == "kusama":
        url = "wss://kusama-rpc.polkadot.io"
        chain = "kusama"
    elif chain == "westend":
        url = "wss://westend-rpc.polkadot.io"
        chain = "westend"
    else:
        return -1

    os.environ["SAS_SUBSTRATE_WS_URL"] = url
    proc = subprocess.Popen(
        ["yarn", "dev"],
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT
    )

    print('Sidecar is loading in development mode...')
    time.sleep(30)

    print('Running `yarn`, and checking/loading cache...')
    run_process(["yarn"])

    print('Running e2e tests for {}'.format(chain))
    res = run_process(["yarn", "test:e2e-tests", "--chain", chain])

    print('Killing all processes...')
    for child in psutil.Process(proc.pid).children(recursive=True):
        child.kill()
    proc.kill()

    print(res)
    return "PASS" in res


def main():
    arg_parser = argparse.ArgumentParser(
        description='Run tests against live chains.'
    )
    arg_parser.add_argument(
        '--chain',
        type=str,
        default='',
        help='Only run the tests against this chain',
        choices=['polkadot', 'kusama', 'westend']
    )

    args = arg_parser.parse_args()

    # Run tests against a specific chain only:
    if args.chain:
        if run_chain_test(args.chain):
            return 0
        else:
            return -1
    # Run tests against all chains of interest to us:
    else:
        if run_chain_test("polkadot") and run_chain_test("kusama") and run_chain_test("westend"):
            return 0
        else:
            return -1


if __name__ == "__main__":
    main()

File: scripts/runChainTests.ts
import { ArgumentParser, Namespace } from 'argparse';

import { config, defaultSasBuildOpts } from './config';
import {
	killAll,
	launchProcess,
	setLogLevel,
	setWsUrl,
} from './sidecarScriptApi';
import { ProcsType, StatusCode } from './types';

// Stores all the processes
const procs: ProcsType = {};

/**
 * Launches Sidecar, and if successful it will launch the jest runner. This operation
 * handles killing all the processes after the jest runner is done.
 *
 * @param chain The chain in which to target the e2e tests too.
 */
const launchChainTest = async (chain: string): Promise<boolean> => {
	const { wsUrl, SasStartOpts, JestProcOpts } = config[chain];
	const { Success } = StatusCode;

	// Set the ws url env var
	setWsUrl(wsUrl);

	console.log('Launching Sidecar...');
	const sidecarStart = await launchProcess('yarn', procs, SasStartOpts);

	if (sidecarStart === Success) {
		// Sidecar successfully launched, and jest will now get called
		console.log('Launching jest...');
		const jest = await launchProcess('yarn', procs, JestProcOpts);

		if (jest === Success) {
			killAll(procs);
			return true;
		} else {
			killAll(procs);
			return false;
		}
	} else {
		console.error('Error launching sidecar... exiting...');
		killAll(procs);
		process.exit(1);
	}
};

const checkTests = (...args: boolean[]) => {
	const testStatus = args.every((test) => test);

	if (testStatus) {
		console.log('[PASSED] All Tests Passed!');
		process.exit(0);
	} else {
		console.log('[FAILED] Some Tests Failed!');
		process.exit(1);
	}
};

const main = async (args: Namespace): Promise<void> => {
	const { Failed } = StatusCode;

	if (args.log_level) {
		setLogLevel(args.log_level);
	}

	// Build sidecar
	console.log('Building Sidecar...');
	const sidecarBuild = await launchProcess('yarn', procs, defaultSasBuildOpts);

	// When sidecar fails to build, we kill all process's and exit
	if (sidecarBuild === Failed) {
		console.error('Sidecar failed to build, exiting...');
		// Kill all processes
		killAll(procs);
		// Exit program
		process.exit();
	}

	if (args.chain) {
		const selectedChain = await launchChainTest(args.chain);

		checkTests(selectedChain);
	} else {
		// Test the e2e tests against polkadot
		const polkadotTest = await launchChainTest('polkadot');

		// Test the e2e tests against kusama
		const kusamaTest = await launchChainTest('kusama');

		// Test the e2e tests against westend
		const westendTest = await launchChainTest('westend');

		// Test the e2e tests against statemine
		const statemineTest = await launchChainTest('statemine');

		// Test the e2e tests against statemint
		const statemintTest = await launchChainTest('statemint');

		checkTests(polkadotTest, kusamaTest, westendTest, statemineTest, statemintTest);
	}
};

/**
 * Arg Parser
 */
const parser = new ArgumentParser();

parser.add_argument('--chain', {
	choices: ['polkadot', 'kusama', 'westend', 'statemine', 'statemint'],
});
parser.add_argument('--log-level', {
	choices: ['error', 'warn', 'info', 'http', 'verbose', 'debug', 'silly'],
});

const args = parser.parse_args() as Namespace;

/**
 * Signal interrupt
 */
process.on('SIGINT', function () {
	console.log('Caught interrupt signal');
	killAll(procs);
	process.exit();
});

/**
 * Signal hangup terminal
 */
process.on('SIGHUP', function () {
	console.log('Caught terminal termination');
	killAll(procs);
	process.exit();
});

main(args).finally(() => process.exit());
